### PR TITLE
Handling cases where non-existing variables are being referenced in URL Map

### LINKED ISSUE #20147: If UrlMap specifies an invalid field, then you can't open the edit content screen
We have an issue in the `URLMapAPIImpl`.  We call this from the the `CMSUrlUtil` all the time.  If a dotCMS installation has an invalid UrlMap - where a user has specified a field that does not exist in their pattern, this error gets thrown.

To reproduce:

create a content type with a URLMap Pattern that specifies a non-valid field, e.g.

```/myUrlMa/{thisFieldDoesNotExist}```

and and then try to edit any content.  

You will get this in the logs:

```
java.lang.NullPointerException
	at com.dotmarketing.cms.urlmap.URLMapAPIImpl.buildFields(URLMapAPIImpl.java:225)
	at com.dotmarketing.cms.urlmap.URLMapAPIImpl.buildContentQuery(URLMapAPIImpl.java:313)
	at com.dotmarketing.cms.urlmap.URLMapAPIImpl.getContentlet(URLMapAPIImpl.java:247)
	at com.dotmarketing.cms.urlmap.URLMapAPIImpl.getContentlet(URLMapAPIImpl.java:95)
	at com.dotmarketing.cms.urlmap.URLMapAPIImpl.isUrlPattern(URLMapAPIImpl.java:53)
	at com.dotmarketing.filters.CMSUrlUtil.isPageAsset(CMSUrlUtil.java:206)
	at com.dotmarketing.filters.CMSUrlUtil.amISomething(CMSUrlUtil.java:396)
	at org.apache.jsp.html.portlet.ext.contentlet.field.tiny_005fmce_005fconfig_jsp._jspService(tiny_005fmce_005fconfig_jsp.java:171)
```

### PR DESCRIPTION
**Pull request by jcastro-dotcms**
There are situations in which customers update their URL Maps with variables that do not exist in the respective Content Type yet. This is causing the application to fail in several key features. In the URL Map API, if a field doesn't exist in the specific Content Type, the application needs to ignore it, and move on.

Also, the `matchingUrlPattern()` method was removed because it's being called twice when verifying if an incoming URL belongs to a URL Map. The very same logic is being called right at the beginning of the `getContentlet()` method too.

Adding missing Javadoc to several methods, and some simple refactoring.

### CODE REVIEW

**Review by jdotcms — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]
looks good but it need tests

**Review by dsilvam — CHANGES_REQUESTED** [review]
Missing tests

File: dotCMS/src/main/java/com/dotmarketing/util/RegEX.java [hunk 1]
```diff
@@ -164,32 +164,55 @@ public static String replaceAll(String original,String substitution,String regEx
...
+     * @param text     The incoming String.
+     * @param regEx    The Regular Expression.
+     * @param isUrlMap Flag indicating if the operation is specific for a URL Map.
+     *
+     * @return The list of {@link RegExMatch} objects.
+     *
+     * @throws DotRuntimeException An error occurred with the specified RegEx.
+     */
+    private static List<RegExMatch> find(final String text, final String regEx, final boolean isUrlMap) throws DotRuntimeException{
+		final RegEX i = getInstance();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/util/RegEX.java:194]: better name here

File: dotCMS/src/main/java/com/dotmarketing/util/RegEX.java [hunk 2]
```diff
@@ -164,32 +164,55 @@ public static String replaceAll(String original,String substitution,String regEx
...
-			Logger.error(RegEX.class, "Unable to compile pattern for regex", e);
-			throw new DotRuntimeException("Unable to compile pattern for regex",e);
+		} catch (final MalformedPatternException e) {
+			final String errorMsg = String.format("Unable to compile pattern for regex with Text = [ %s ], regEx = [ " +
+					"%s ]: %s", text, regEx, e.getMessage());
+			Logger.error(RegEX.class, errorMsg, e);
+			throw new DotRuntimeException(errorMsg, e);
 		}
-		List<RegExMatch> res = new ArrayList<RegExMatch>();
+		final List<RegExMatch> res = new ArrayList<>();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/util/RegEX.java:205]: better name here

**Review by jdotcms — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Pull request merged**